src/internal/consistenthashing/etcd.go

```diff
@@ -0,0 +1,331 @@
+		members: []member{localMember},
+		node: node{
+			member: localMember,
+			locks:  map[string]lockInfo{},
+		},
+		prefix: prefix,
+	}
+}
+
+// createNode creates a lease, inserts itself as a key to etcd, keeps the lease alive in the background.
+func (ring *Ring) createNode(ctx context.Context, cancel context.CancelFunc, col collection.EtcdCollection, id string) error {
+	if err := col.Claim(ctx, id, &types.BoolValue{Value: true},
+		func(ctx context.Context) error {
+			// keep the lease alive until the context is canceled.
+			<-ctx.Done()
+			return errors.EnsureStack(ctx.Err())
+		}); err != nil {
+		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+			return nil
+		}
+		cancel()
+		ring.logger.WithFields(ring.memberLogFields()).Errorf("failed to renew etcd lease: %v", err)
+		return errors.EnsureStack(err)
+	}
+	return nil
+}
+
+// watch watches etcd keys with prefix Ring.prefix to determine if a node has been added or removed.
+// When a node is added, the ring calls rebalance to release any locks that no longer associate to its node. Rebalance
+// on delete happens by default since each member attempts to retrieve all locks. When a member is deleted,
+// the pending calls to Lock by each member will go through on the nodes that associates to given lock.
+func (ring *Ring) watch(ctx context.Context, cancel context.CancelFunc, col collection.EtcdCollection) error {
+	if err := col.ReadOnly(ctx).WatchF(func(event *watch.Event) error {
+		ring.stateLock.Lock()
+		defer ring.stateLock.Unlock()
+		nodeKey := string(event.Key)
+		switch event.Type {
+		case watch.EventDelete:
+			ring.removeById(ring.etcdToMember(nodeKey))
+		case watch.EventPut:
+			ring.insertById(ring.etcdToMember(nodeKey))
+			// Need to release locks that no longer associate to the ring's node.
+			if err := ring.rebalance(); err != nil {
+				ring.logger.WithField("ring", ring.prefix).Errorf("failed rebalancing: %v", err)
+				return err
+			}
+		}
+		return nil
+	}); err != nil {
+		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+			ring.logger.WithFields(ring.memberLogFields()).Infof("shutting down ring")
+			return nil
+		}
+		cancel()
+		ring.logger.WithFields(ring.memberLogFields()).Errorf("failed watch: %v", err)
+		return errors.EnsureStack(err)
+	}
+	return nil
+}
+
+// etcdToMember converts an etcd key in a ring's collection to a node ID.
+func (ring *Ring) etcdToMember(key string) string {
+	return path.Base(key)
+}
+
+// keyWithRingPrefix prepends the ring prefix to a key to prevent key collisions with other rings.
+func (ring *Ring) keyWithRingPrefix(key string) string {
+	return path.Join(ring.prefix, key)
+}
+
+// rebalance iterates through held locks and releases locks that no longer associate to a ring's node.
+func (ring *Ring) rebalance() error {
+	var errs error
+	for key, lockInfo := range ring.node.locks {
+		if ring.get(key).Id == ring.node.Id || lockInfo.lock == nil {
```

Comment:
When would the lock be nil if there is a lock info?

Reply:
this is a holdover from before when we were setting the lockInfo ahead of time and storing a cancel, this is not needed anymore.